Extensions/PrimitiveDrawing/shapepainterruntimeobject.ts

```diff
@@ -438,23 +450,294 @@ namespace gdjs {
         return;
       }
       super.setX(x);
-      this._renderer.updateXPosition();
+      this._renderer.updatePositionX();
     }
 
     setY(y: float): void {
       if (y === this.y) {
         return;
       }
       super.setY(y);
-      this._renderer.updateYPosition();
+      this._renderer.updatePositionY();
+    }
+
+    setAngle(angle: float): void {
+      if (angle === this.angle) {
+        return;
+      }
+      super.setAngle(angle);
+      this._renderer.updateAngle();
+    }
+
+    /**
+     * The center of rotation is defined relatively
+     * to the drawing origin (the object position).
+     * This avoid the center to move on the drawing
+     * when new shapes push the bounds.
+     *
+     * When no custom center is defined, it will move
+     * to stay at the center of the drawable bounds.
+     *
+     * @param x coordinate of the custom center
+     * @param y coordinate of the custom center
+     */
+    setRotationCenter(x: float, y: float): void {
+      if (!this._customCenter) {
+        this._customCenter = [0, 0];
+      }
+      this._customCenter[0] = x;
+      this._customCenter[1] = y;
+    }
+
+    /**
+     * @returns The center X relatively to the drawing origin
+     * (whereas `getCenterX()` is relative to the top left drawable bound and scaled).
+     */
+    getRotationCenterX(): float {
+      return this._customCenter
+        ? this._customCenter[0]
+        : this._renderer.getUnscaledWidth() / 2 - this._renderer.getOriginX();
+    }
+
+    /**
+     * @returns The center Y relatively to the drawing origin
+     * (whereas `getCenterY()` is relative to the top left drawable bound and scaled).
+     */
+    getRotationCenterY(): float {
+      return this._customCenter
+        ? this._customCenter[1]
+        : this._renderer.getUnscaledHeight() / 2 - this._renderer.getOriginY();
+    }
+
+    getCenterX(): float {
+      if (!this._customCenter) {
+        return super.getCenterX();
+      }
+      let centerX =
+        this._customCenter[0] +
+        (this.getX() - this.getDrawableX()) / Math.abs(this._scaleX);
+      return centerX * Math.abs(this._scaleX);
```

Comment:
Can we simplify in:

```suggestion
        return this._customCenter[0] * Math.abs(this._scaleX) + this.getX() - this.getDrawableX();
```
?